Make the archives created on Unix have the read permissions by default

derwasp opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/17912

## Proposed changes

* When creating a `ZipArchiveEntry` in the Unix environment, always set the default file permissions to be `644` and directory permissions to be `755` in order for the archives to be usable after running an `unzip` command on them.
* Always override the default permissions of the `ZipArchiveEntry` when the archive is created from the actual files and not from memory, where the permissions are not available.

## Background

Recently we hit an issue where our customers saw that the artifact files in the zip archives that we produce have the permissions set to `000` if our system runs on Unix machines. When the `unzip` command is used, these permissions translate to the actual files also having the same permissions.  
This was an inconvenience to them as if the same system is run on Windows, the files would have the `644` permissions. The customers wanted this to have the same behavior.

## Causes

Based on how I see it, the zip archives have two fields that can be responsible for the file permissions.  

### Version made by

One of them is `Version made by`. (See the `Central directory file header` section in the [wiki page](https://en.wikipedia.org/wiki/ZIP_(file_format)) ) This is what the comment says about this field in the code:

```
The upper byte of the "version made by" flag in the central directory header of a zip file represents the
OS of the system on which the zip was created. Any zip created with an OS byte not equal to Windows (0)
or Unix (3) will be treated as equal to the current OS.
The value of 0 more specifically corresponds to the FAT file system while NTFS is assigned a higher value. However
for historical and compatibility reasons, Windows is always assigned a 0 value regardless of file system.
```

So to simplify the story, we can say that we have two possible options here: `0` for `Windows` and `3` for `Unix`.

This field is written here: [ZipArchiveEntry.cs#L517](https://github.com/dotnet/runtime/blob/83adfae6a6273d8fb4c69554aa3b1cc7cbf01c71/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L517)

The unzipping software is using this field to determine that the permissions were present in the source system. In other words: we probably need to set the permissions to `644` on the Unix system if the archive was created on Windows, where these permissions don't exist. And this is exactly the behavior with `unzip` that we see.

### External attributes

Another field is `External attributes`. This field is used to store in the file permissions when the archives are created on the Unix systems. Recently, the API that works with files received support for this: [PR](https://github.com/dotnet/runtime/pull/55531).  
As you can see from the PR, the permissions are based on the actual permissions from the files:

```csharp
    static partial void SetExternalAttributes(FileStream fs, ZipArchiveEntry entry)
    {
        Interop.Sys.FileStatus status;
        Interop.CheckIo(Interop.Sys.FStat(fs.SafeFileHandle, out status), fs.Name);

        entry.ExternalAttributes |= status.Mode << 16;
    }
```

So now from the perspective of the unzipping software if we have a `Version made by` set to `3` (meaning `Unix`), we need to read the `External attributes` and apply the corresponding permissions to the files. And if `Version made by` is set to `0` (meaning `Windows`) we need determine the permissions ourself and probably set it to `644`.

### The issue

If the file API is used, then everything already works: the permissions are preserved and when the unpacking software is going to work with the archive it would create the files with same permissions as before.  
The caviat is that when we are creating a new `ZipArchiveEntry` from memory on the Unix machines. In this case, there's no file to begin with, but this new `ZipArchiveEntry` can still be unzipped using other software, such as `unzip`, which would follow the logic based on the `Version made by` flag.  

When this happens, currently, we would create an archive which, when unzipped, would result in files having `000` permissions if our zipping logic runs on `Windows` and `644` (probably depends on the implementation of the unzipping software) if our zipping logic runs on `Unix`.
As noted by @tmds we should also set the permissions to `755` for the directories as in Unix the `execute` permissions is required to `cd` into the folder.

Even if this is not techincally a bug, at least, this behavior is not consistent across the two platforms from the users point of view.

A minimal repro looks like this:

```csharp
using System.IO.Compression;

using var fileStream = new FileStream(@"test.zip", FileMode.Create);
using var archive = new ZipArchive(fileStream, ZipArchiveMode.Create, true);

ZipArchiveEntry demoFile = archive.CreateEntry("inmemory_file.txt");

using var entryStream = demoFile.Open();
using var streamWriter = new StreamWriter(entryStream);
streamWriter.Write("Hello zip!");

Console.WriteLine("Done.");

```

After running this on a `Unix` machine and unizipping the result with `unzip test.zip` you can do an `ls -la` and you would see something like this:

```
 xxxx@xxxx  ~/Projects/ConsoleApp1/ConsoleApp1/bin/Debug/net6.0  ls -la
drwxr-xr-x  10 derwasp  staff     320 Jun 14 14:48 .
drwxr-xr-x   3 derwasp  staff      96 Jun 13 16:42 ..
-rwxr-xr-x   1 derwasp  staff  132208 Jun 14 14:48 ConsoleApp1
-rw-r--r--   1 derwasp  staff     403 Jun 14 14:48 ConsoleApp1.deps.json
-rw-r--r--   1 derwasp  staff    5120 Jun 14 14:48 ConsoleApp1.dll
-rw-r--r--   1 derwasp  staff   10480 Jun 14 14:48 ConsoleApp1.pdb
-rw-r--r--   1 derwasp  staff     139 Jun 14 14:48 ConsoleApp1.runtimeconfig.json
----------   1 derwasp  staff      10 Jun 14 14:48 inmemory_file.txt
drwxr-xr-x   3 derwasp  staff      96 Jun 14 14:48 ref
-rw-r--r--   1 derwasp  staff     150 Jun 14 14:48 test.zip
```

And you can see that the file permissions are set to `000`.


## Links 
The original comment is here: https://github.com/dotnet/runtime/issues/17912#issuecomment-1155158903

derwasp committed:
Set the ExternalAttributes to the values from the file.

We need to set the external attributes to the *exact* values that the files have. Otherwise if we create a ZipArchive entry and set something there this would get added to the file values and this is not what we want.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70735) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70735) before we can accept your contribution.<br/><br/>:x: derwasp [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=70735)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=70735) it.</sub>

danmoseley commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.Unix.cs:
```diff
@@ -7,6 +7,13 @@ public partial class ZipArchiveEntry
+        /// For Unix the default permissions are an equivalent of (Convert.ToInt32("644", 8) &lt;&lt; 16)
+        /// where 644 are RW-R-R permissions.
+        /// </summary>
+        internal const int DefaultEntryPermissions = 420 << 16;
```
the `<< 16` is to match where they're stored in ExternalAttributes, right? The perms are actually 420==0644 so the name of this is confusing to me. Can the shift happen at the point its set on ExternalAttributes?

danmoseley replied:
Or, since you are just going to set them not or them, you could call this `DefaultExternalAttributes`

derwasp replied:
I agree, I renamed the const to be `DefaultExternalAttributes` as you suggested.

derwasp replied:
This has since then been renamed back as now I am using the enum in the constants and I am shifting it right where we set the attributes.

tmds commented:
Is something needed to handle directory permissions?

derwasp commented:
@tmds I used my example from the issue comment and modified it to have a line that adds an archive folder like so:

```csharp
using System.IO.Compression;
using ConsoleApp1;

using var fileStream = new FileStream(@"test.zip", FileMode.Create);
using var archive = new ZipArchive(fileStream, ZipArchiveMode.Create, true);

archive.CreateEntry("folder/").EnsureReadPermissions();
ZipArchiveEntry demoFile = archive.CreateEntry("inmemory_file.txt").EnsureReadPermissions();

using var entryStream = demoFile.Open();
using var streamWriter = new StreamWriter(entryStream);
streamWriter.Write("Hello zip!");

Console.WriteLine("Done.");
```

After running this and unzipping this is the result:

```
derwasp@xxxxx  ~/RiderProjects/ConsoleApp1/ConsoleApp1/bin/Debug/net6.0/out  ls -la
    drwxr-xr-x   4 derwasp  staff  128 Jun 22 09:23 .
    drwxr-xr-x  11 derwasp  staff  352 Jun 22 09:23 ..
    drw-r--r--   2 derwasp  staff   64 Jun 22 09:22 folder           <-- permissions
    -rw-r--r--   1 derwasp  staff   10 Jun 22 09:22 inmemory_file.txt

```

And this is the result if I don't see the permissions to the folder:

```
derwasp@xxxxx  ~/RiderProjects/ConsoleApp1/ConsoleApp1/bin/Debug/net6.0/out  ls -la
    drwxr-xr-x   4 derwasp  staff  128 Jun 22 09:23 .
    drwxr-xr-x  11 derwasp  staff  352 Jun 22 09:23 ..
    d---------   2 derwasp  staff   64 Jun 22 09:23 folder           <-- permissions
    -rw-r--r--   1 derwasp  staff   10 Jun 22 09:23 inmemory_file.txt
```

So this logic should fix the permissions for both, files and folders.

derwasp commented:
@tmds oh this is a very good catch, I didn't know that `r` is not enough for the directories!

tmds commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -113,7 +113,10 @@ internal ZipArchiveEntry(ZipArchive archive, string entryName)
-            _externalFileAttr = 0;
+            _externalFileAttr = entryName.EndsWith(Path.DirectorySeparatorChar) || entryName.EndsWith(Path.AltDirectorySeparatorChar)
+                                ? DefaultDirectoryExternalAttributes
+                                : DefaultFileExternalAttributes;
```
Note that this will also affect the empty directory entries created from file system:
https://github.com/dotnet/runtime/blob/fc813dd6839cf327bdd994039532bfcbbef7d4c9/src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.Create.cs#L412-L414

derwasp replied:
I remember working with zip files before and having different directory separators. I just tried a small app and it seems like no translation happens there, which is sort of expected in case you want to have the specific separators inside.

This is the output after running the app from the spoiler:
```
folder/file.txt
folder\file.txt
Done.
```

<details>
  <summary>Program.cs</summary>
  

 ```csharp
using System.IO.Compression;


await using (var fileStream = new FileStream(@"test.zip", FileMode.Create))
using (var archive = new ZipArchive(fileStream, ZipArchiveMode.Create, true))
{
    await archive.CreateSomeEntry("folder/file.txt");
    await archive.CreateSomeEntry("folder\\file.txt");
}

await using (var fileStream = new FileStream(@"test.zip", FileMode.Open))
using (var archive = new ZipArchive(fileStream, ZipArchiveMode.Read, true))
{
    foreach (var e in archive.Entries)
    {
        Console.WriteLine(e.FullName);
    }
}


Console.WriteLine("Done.");

public static class Extensions
{
    public static ZipArchiveEntry EnsureReadPermissions(this ZipArchiveEntry entry, bool folder = false)
    {
        if (!OperatingSystem.IsWindows())
        { 
            int filePermissions = 420 << 16;
            int folderPermissions = 493 << 16;
            
            entry.ExternalAttributes = folder ? folderPermissions : filePermissions;
        }

        return entry;
    }

    public static async Task CreateSomeEntry(this ZipArchive archive, string path)
    {
        ZipArchiveEntry demoFile = archive.CreateEntry(path).EnsureReadPermissions();

        await using var entryStream = demoFile.Open();
        await using var streamWriter = new StreamWriter(entryStream);
        await streamWriter.WriteAsync("Hello zip!");
    }
}
  ```
  
</details>

tmds commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.Unix.cs:
```diff
@@ -7,6 +7,20 @@ public partial class ZipArchiveEntry
+        /// For Unix the default external attributes are an equivalent of (Convert.ToInt32("755", 8) &lt;&lt; 16)
+        /// where 755 are rwxr-xr-x unix file permissions.
+        /// </summary>
+        internal const uint DefaultDirectoryExternalAttributes = 493 << 16;
```
@carlossanlop similar to the change here, we probably need a default for directories also in the Tar implementation.

tmds commented:
Looking at this PR made me realize Zip doesn't store/restore directory permissions.
I've created a ticket for that: https://github.com/dotnet/runtime/issues/71140.

tmds commented:
@derwasp can you add a test that verifies the default attributes are used?

danmoseley commented on src/libraries/System.IO.Compression/tests/ZipArchive/zip_CreateTests.Unix.cs:
```diff
@@ -0,0 +1,65 @@
+                await CreateArchive();
+                await UnpackArchive();
+
+                EnsureFilePermissions(Path.Combine(outputFolder, nestedFolder, file), "644");
```
and check folder is 755?

derwasp replied:
Ideally, yes. But with the current unzipping implementation, we can't do that: the folders are just created using [`Directory.CreateDirectory`](https://cs.github.com/dotnet/runtime/blob/5877e8b713742b6d80bd1aa9819094be029e3e1f/src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchiveEntry.Extract.cs#L110) . When `unzip` is used, the permissions are fully respected of course.
IMO changing the unzipping logic fits this issue better:
 - https://github.com/dotnet/runtime/issues/71140

tmds replied:
We can reduce what these tests are doing to: creating a `ZipArchiveEntry`, and verifying `ExternalAttributes` has the expected value.
That will also work for directories.

tmds replied:
The existing `UnixExtractSetsFilePermissionsFromExternalAttributes` verifies `ExternalAttributes` gets stored and restored.

derwasp replied:
@tmds I made a very simple test instead of the one I had previously. Is this what you had in mind?

tmds commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.Unix.cs:
```diff
@@ -7,6 +7,20 @@ public partial class ZipArchiveEntry
+        /// For Unix the default external attributes are an equivalent of (Convert.ToInt32("644", 8) &lt;&lt; 16)
+        /// where 644 are rw-r--r-- unix file permissions.
+        /// </summary>
+        internal const uint DefaultFileExternalAttributes = 420 << 16;
```
@derwasp can you use the `UnixFileMode` that got just merged, and use the same style as:

https://github.com/dotnet/runtime/blob/b869f40132d9f152bdc46b5a760953579843bae9/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHelpers.cs#L23-L24

@carlossanlop @eerhardt @danmoseley should we try to put these consts in a file common to the `Tar` and `Zip` implementations to ensure they use the same values?

derwasp replied:
@tmds Yeah, I just saw that your PR went in, awesome timing! Looking at the API right now.

tmds commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.Windows.cs:
```diff
@@ -7,6 +7,18 @@ public partial class ZipArchiveEntry
+        /// The default external file attributes are used to support zip archives on multiple platforms.
+        /// Since Windows doesn't use file permissions, there's no default value needed.
+        /// </summary>
+        internal const uint DefaultFileExternalAttributes = 0;
```
I'm curious. What happens when you extract the `ZipVersionMadeByPlatform.Windows` file on Unix using the system tools. What permissions do you get?

eerhardt replied:
https://github.com/dotnet/runtime/blob/ddc4f95432facf2deebfa85bad7143469dca0129/src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchiveEntry.Extract.Unix.cs#L20-L26

derwasp replied:
@tmds it took me a while to test this: was trying to understand how to set up the dev environment on my old windows laptop. 

Here's the result unpacked on a mac machine:
```
 derwasp@xxxxx  ~/Downloads/ziptest > unzip archive_created_on_win.zip -d win_arch
Archive:  archive_created_on_win.zip
   creating: win_arch/folder/
  inflating: win_arch/folder/inmemory_file.txt
 derwasp@xxxxx  ~/Downloads/ziptest > cd win_arch
 derwasp@xxxxx  ~/Downloads/ziptest/win_arch > ls -la
total 0
drwxr-xr-x   3 derwasp  staff   96 Jun 24 22:49 .
drwxr-xr-x  14 derwasp  staff  448 Jun 24 22:49 ..
drwxr-xr-x   3 derwasp  staff   96 Jun 24 13:45 folder
 derwasp@xxxxx  ~/Downloads/ziptest/win_arch > cd folder
 derwasp@xxxxx  ~/Downloads/ziptest/win_arch/folder > ls -la
total 8
drwxr-xr-x  3 derwasp  staff  96 Jun 24 13:45 .
drwxr-xr-x  3 derwasp  staff  96 Jun 24 22:49 ..
-rw-r--r--  1 derwasp  staff  56 Jun 24 13:45 inmemory_file.txt
```

I also verified that the default values are not set like so:

```
var arch = "/Users/derwasp/Downloads/ziptest/archive_created_on_win.zip";

await using (var fileStream = new FileStream(arch, FileMode.Open))
using (var archive = new ZipArchive(fileStream, ZipArchiveMode.Read, true))
{
    foreach (var e in archive.Entries)
    {
        Console.WriteLine($"{e.FullName}, \t\t attributes = {(e.ExternalAttributes >> 16)}");

    }
}
```

And this is the output:

```
folder/,                 attributes = 0
folder/inmemory_file.txt,                attributes = 0
```

So the `unzip` utility does it the way that we expected: `644` for the files and `755` for the folders.

eerhardt commented on src/libraries/System.IO.Compression/tests/System.IO.Compression.Tests.csproj:
```diff
@@ -1,5 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
+    <EnableLibraryImportGenerator>true</EnableLibraryImportGenerator>
```
Why is this necessary?

derwasp replied:
This was necessary before I switched to the new API: it made the interop methods work. It is probably no longer necessary. I will remove it once I get my Codespaces instance to behave and test the new version as I can't compile it right now.

eerhardt commented on src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.Unix.cs:
```diff
@@ -10,7 +10,7 @@ public static partial class ZipFileExtensions
             Interop.CheckIo(Interop.Sys.FStat(fs.SafeFileHandle, out status), fs.Name);
 
-            entry.ExternalAttributes |= status.Mode << 16;
+            entry.ExternalAttributes = status.Mode << 16;
```
The ExternalAttributes store more things than just permissions, right? That's why we are bit-shifting the Mode 16.

So this is overwriting all the other bits already set in ExternalAttributes (if any).

derwasp replied:
This seems correct, yes. My logic was this:

1) The default external attributes are `0` right now and I am going to change them to be `644` for files and `755` for folders (also `<< 16`) when running on a Unix system.
2) Since the default values are going to be set, we need to ensure that when somebody is using the API that gets the actual permissions from the files, gets the permissions that the files have.
3) This method, `SetExternalAttributes` is private (because it doesn't have a modifier on it) and is only used in `DoCreateEntryFromFile`. And it's called right after we create an entry.

So to me it seems like we are not overwriting anything there when we don't have to. Am I missing something?

derwasp replied:
@eerhardt  I like the idea.
What would be your suggestion regarding making this const accessible in the `System.IO.Compression.ZipFile` assembly? Right now it's defined as internal and resides in `System.IO.Compression`.
Should we use `InternalsVisibleTo`? Or should we just make it public? Or any other options?

eerhardt replied:
I would make an internal shared source file that is compiled into both assemblies.

derwasp replied:
@eerhardt Can you please check if this is how you wanted it?

eerhardt replied:
> so I don't find the assert necessary

My intention with the assert is to catch the case in the future where some new code gets added that is setting other pieces of the ExternalAttributes field. If there is ever a case where other pieces of ExternalAttributes is set, and then this method is called, we will be overwriting that information. The assert is there to alert that future developer that this code may need to change - for example, in that case it could just clear the permission bits first, and then `|= status.Mode << 16`

> the name SetExternalAttributes

Maybe a better name for this method would be `SetPermissions` or `SetFileMode`.

derwasp replied:
Do you guys want me to change the name or should we keep it for now? Im okay with both.
Also, I think we might keep the assert: it doesn't really break anything. I like being extra careful.

derwasp committed:
Add an assert of te SetExternalAttributes

derwasp committed:
Move the constants into separate files

derwasp committed:
Add a test verifying the new behavior

derwasp committed:
Use the new default permissions on the ZipArchiveEntry

derwasp committed:
Move the zip constants to the common directory

Move the new zip constants to the common folder and use the constants files in both IO.Compression and IO.Compression.ZipFile

tmds commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -115,7 +117,11 @@ internal ZipArchiveEntry(ZipArchive archive, string entryName)
             _compressedSize = 0; // we don't know these yet
             _uncompressedSize = 0;
-            _externalFileAttr = 0;
+            var defaultEntryPermissions = entryName.EndsWith(Path.DirectorySeparatorChar) || entryName.EndsWith(Path.AltDirectorySeparatorChar)
```
@eerhardt if you're adding stuff from file system, you go through a function that replaces `Path.DirectorySeparatorChar` and `Path.AltDirectorySeparatorChar` with `/`:

https://github.com/dotnet/runtime/blob/0f75a0269ce14d5838590cb4c634ed685c044d62/src/libraries/Common/src/System/IO/Archiving.Utils.cs#L41-L49

On Linux, `\` is a filename char, and we're replacing it there. Maybe we shouldn't?

For entries created using `ZipArchive.CreateEntry` we're not performing any substitutions.
Maybe we assume the user of this API uses proper separators?
Should we limit to the `Path.DirectorySeparatorChar` for this check?

eerhardt commented on src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.Unix.cs:
```diff
@@ -1,16 +1,28 @@
+            // This assert ensures that nothing else sets values different to the default before
+            // overwriting it with the data read from the files
+            Debug.Assert(entry.ExternalAttributes == ((uint)DefaultFileEntryPermissions << 16)
+                         || entry.ExternalAttributes == ((uint)DefaultDirectoryEntryPermissions << 16));
```
This is only ever called for files (i.e. it takes a `FileStream`), so we shouldn't need to check for `DefaultDirectoryEntryPermissions`.

derwasp replied:
Removed the `DefaultDirectoryEntryPermissions` check. I wanted to first change this method to also work with the directories but then decided not to do it in this PR.

eerhardt commented on src/libraries/System.IO.Compression.ZipFile/src/System.IO.Compression.ZipFile.csproj:
```diff
@@ -18,13 +18,15 @@
   </ItemGroup>
   <!-- Windows specific files -->
   <ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'windows'">
+    <Compile Include="$(CommonPath)System\IO\Compression\ZipArchiveEntryConstants.Windows.cs" />
```
This is never used on Windows, correct? I think we can delete this line and just include `ZipArchiveEntryConstants.Unix.cs` on non-Windows

```suggestion
```

derwasp replied:
Oh, good point, thank you @eerhardt

derwasp commented on src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.Unix.cs:
```diff
@@ -1,16 +1,28 @@
+            // SetExternalAttributes is only used to overwrite the default values when reading on the unix systems
+            // This assert ensures that nothing else sets values different to the default before
+            // overwriting it with the data read from the files
+            Debug.Assert(entry.ExternalAttributes == ((uint)DefaultFileEntryPermissions << 16)
+                         || entry.ExternalAttributes == ((uint)DefaultDirectoryEntryPermissions << 16));
```
```suggestion
            Debug.Assert(entry.ExternalAttributes == ((uint)DefaultFileEntryPermissions << 16));
```

tmds reviewed (approved):
lgtm, thanks @derwasp!

adamsitnik commented:
/azp run runtime-extra-platforms

eerhardt reviewed (approved):
LGTM - thank you @derwasp for the contribution!

adamsitnik reviewed (approved):
LGTM, thank you for your contribution @derwasp !

derwasp commented:
Thank you for merging this in and for your help @tmds @eerhardt @adamsitnik :)

danmoseley commented:
@derwasp we'd welcome another contribution if you're interested!
https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aup-for-grabs
